### PR TITLE
IDE+Evaluator: converge multiple type relation checks into a single sema request, NFC

### PR DESCRIPTION
**Pull request by nkcsgexi**
From libIDE, the utility functions will invoke the request in the implementation.

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: lib/Sema/CSGen.cpp [hunk 1]
```diff
@@ -3769,31 +3769,14 @@ bool swift::areGenericRequirementsSatisfied(
   return CS.solveSingle().hasValue();
 }
 
-static bool canSatisfy(Type type1, Type type2, bool openArchetypes,
+bool swift::canSatisfy(Type type1, Type type2, bool openArchetypes,
```
Comment (slavapestov) [inline lib/Sema/CSGen.cpp]: Can this be moved to lib/IDE as well?

Comment (nkcsgexi) [reply lib/Sema/CSGen.cpp]: It's quite difficult. The implementation of `canSatisfy()` invokes `TypeChecker::typesSatisfyConstraint` which is not part of sema APIs we can call from libIDE.

File: lib/Sema/CSGen.cpp [hunk 2]
```diff
@@ -3769,31 +3769,14 @@ bool swift::areGenericRequirementsSatisfied(
...
                        ConstraintKind kind, DeclContext *dc) {
   std::unique_ptr<TypeChecker> CreatedTC;
   auto &TC = TypeChecker::createForContext(dc->getASTContext());
   return TC.typesSatisfyConstraint(type1, type2, openArchetypes, kind, dc,
                                    /*unwrappedIUO=*/nullptr);
 }
 
-bool swift::canPossiblyEqual(Type T1, Type T2, DeclContext &DC) {
```
Comment (slavapestov) [inline lib/Sema/CSGen.cpp]: Is this any different from `TypeBase::isBindableTo()`?

File: lib/Sema/CSGen.cpp [hunk 3]
```diff
@@ -3769,31 +3769,14 @@ bool swift::areGenericRequirementsSatisfied(
...
-  return canSatisfy(T1, T2, true, ConstraintKind::Bind, &DC);
-}
-
-bool swift::canPossiblyConvertTo(Type T1, Type T2, DeclContext &DC) {
```
Comment (slavapestov) [inline lib/Sema/CSGen.cpp]: This is only used in one place, which has a FIXME comment noting that it's wrong :)

File: lib/Sema/CSGen.cpp [hunk 4]
```diff
@@ -3769,31 +3769,14 @@ bool swift::areGenericRequirementsSatisfied(
...
-  return canSatisfy(T1, T2, true, ConstraintKind::Conversion, &DC);
-}
-
-bool swift::isEqual(Type T1, Type T2, DeclContext &DC) {
```
Comment (slavapestov) [inline lib/Sema/CSGen.cpp]: This seems unnecessary

File: lib/Sema/CSGen.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/Sema/CSGen.cpp]: Sorry, I meant move it to IDETypeChecking.cpp

**Comment by slavapestov** [comment]
My comments are just for possible followups. I think moving this code to IDETypeChecking is a good start, but ideally we'd eliminate these as much as possible. They're quite expensive because they create a new ConstraintSystem on each call, and other than checking for conversion constraints, the other queries can be answered in a cheaper manner, I think.

File: lib/Sema/CSGen.cpp (hunk 3 above)
Comment (nkcsgexi) [reply lib/Sema/CSGen.cpp]: Notice this function has been removed altogether.

File: lib/IDE/IDETypeChecking.cpp [hunk 5]
```diff
@@ -327,7 +327,7 @@ struct SynthesizedExtensionAnalyzer::Implementation {
         case RequirementKind::Superclass:
           // FIXME: This could be more accurate; check
           // conformance instead of subtyping
-          if (!canPossiblyConvertTo(First, Second, *DC))
+          if (!isConvertibleTo(First, Second, /*openArchetypes=*/true, *DC))
```
Comment (slavapestov) [inline lib/IDE/IDETypeChecking.cpp]: for a Superclass constraint, you want to check isBindableToSuperclassOf() (that's like the "open archetypes" version), followed by isExactSuperclassOf().

For conformance, we'll need to cook something up. Basically a type parameter "can potentially" conform to anything; for a concrete type you call back to conformsToProtocol().

File: lib/Sema/CSGen.cpp (hunk 3 above)
Comment (slavapestov) [reply lib/Sema/CSGen.cpp]: The FIXME still applies. I added a comment there.

**Comment by nkcsgexi** [comment]
@slavapestov Thank you for taking a look! These changes are mostly about refactoring the existing interaction between libSema and libIDE into the request evaluator model. Enhancing the performance of these actual implementations could come as separate PRs that address the IDE's performance concerns.

**Pull request merged**